### ISSUE TITLE
Pruning a species and adding it to the core at the same time

### ISSUE THREAD

**Issue by chiyw13** [ReactionMechanismGenerator/RMG-Py#1122]
While testing pruning on the AcceleratorTools branch (though this may also apply to Master) I set maximum edge size to 1000 (requiring lots of pruning) and had two reactor systems.

It pruned species 7480 to get below the 1000 limit, but reaction system 2 decided that was the highest flux and added it to the core. This (unsurprisingly) led to an error.

```
Conducting simulation of reaction system 1...
initializing surface ...
surface initialization complete
Moving reaction O2(2) + C=CC=O(6705) <=> [O]C=CCO[O](7833) from surface to core
initializing surface ...
removing disconnected species from surface: [O]C=CCO[O]
surface initialization complete
Surface has 0 Species and 0 Reactions
At time 5.0000e+00 s, reached target termination time.
Making seed mechanism...

Conducting simulation of reaction system 2...
initializing surface ...
surface initialization complete
Moving reaction O2(2) + C=CC=O(6705) <=> [O]C=CCO[O](7833) from surface to core
initializing surface ...
removing disconnected species from surface: [O]C=CCO[O]
surface initialization complete
Surface has 0 Species and 0 Reactions
At time 9.7630e-02 s, species C=CC([O])O(7480) exceeded the minimum rate for moving to model core
    Characteristic rate: 1.1866e+00 mol/m^3*s
    Dynamics number:  2.5599e+00
    C=CC([O])O(7480) rate: 5.9647e-01 mol/m^3*s (0.5027)
Surface has 0 Species and 0 Reactions
At time 5.0000e+00 s, reached target termination time.
Making seed mechanism...

Pruning species C=CC(O)O(6826) to make numEdgeSpecies smaller than maximumEdgeSpecies
Pruning species OC=CCO(6828) to make numEdgeSpecies smaller than maximumEdgeSpecies
Pruning species O=C(O)O(8197) to make numEdgeSpecies smaller than maximumEdgeSpecies
Pruning species [O]C=CCOO(7826) to make numEdgeSpecies smaller than maximumEdgeSpecies
Pruning species C=CC[O](8242) to make numEdgeSpecies smaller than maximumEdgeSpecies
Pruning species [CH2]CC=O(6712) to make numEdgeSpecies smaller than maximumEdgeSpecies
Pruning species [CH]=C(7844) to make numEdgeSpecies smaller than maximumEdgeSpecies
Pruning species C=CC([O])O[O](7479) to make numEdgeSpecies smaller than maximumEdgeSpecies
Pruning species CC=C[O](6711) to make numEdgeSpecies smaller than maximumEdgeSpecies
Pruning species C=C=CO(6704) to make numEdgeSpecies smaller than maximumEdgeSpecies
Pruning species C=CC([O])O(7480) to make numEdgeSpecies smaller than maximumEdgeSpecies
Pruning 859 species whose rate ratios against characteristic rate did not exceed the minimum threshold of 0.02
Pruning species [O]OC[CH]CC(O)C(O)CO[O](8269)                           
Pruning species [O]OC[CH]CC(CO[O])C(O)O(8268)                           
...snip...                                    
Pruning species O=CCCO(8067)                                            
Pruning species OC1[CH]COO1(7469)                                       
Pruning 11 species to obtain an edge size of 1000 species
Pruning species C=CC(O)O(6826)                                          
Pruning species OC=CCO(6828)                                            
Pruning species O=C(O)O(8197)                                           
Pruning species [O]C=CCOO(7826)                                         
Pruning species C=CC[O](8242)                                           
Pruning species [CH2]CC=O(6712)                                         
Pruning species [CH]=C(7844)                                            
Pruning species C=CC([O])O[O](7479)                                     
Pruning species CC=C[O](6711)                                           
Pruning species C=C=CO(6704)                                            
Pruning species C=CC([O])O(7480)                                        

Garbage collector: collected 105161 objects.

Adding species C=CC([O])O(7480) to model core
Generating kinetics for new reactions...

Summary of Model Enlargement
---------------------------------
Added 1 new core species
    C=CC([O])O(7480)
Created 0 new edge species
Added 0 new core reactions
Created 0 new edge reactions

After model enlargement:
    The model core has 75 species and 830 reactions
    The model edge has 1157 species and 1799 reactions


initializing surface ...
surface initialization complete
Moving reaction O2(2) + C=CC=O(6705) <=> [O]C=CCO[O](7833) from surface to core
initializing surface ...
removing disconnected species from surface: [O]C=CCO[O]
surface initialization complete
Surface has 0 Species and 0 Reactions
At time 5.0000e+00 s, reached target termination time.
initializing surface ...
surface initialization complete
At time 5.0000e+00 s, reached target termination time.

The following error occurend on a worker:
7480
Traceback (most recent call last):
  File "/home/chiyw1992/.virtualenvs/rmgpy275/lib/python2.7/site-packages/scoop/_control.py", line 127, in runFuture
    future.resultValue = future.callable(*future.args, **future.kargs)
  File "/shared/apps/python/Python-2.7.5/INSTALL/lib/python2.7/runpy.py", line 240, in run_path
    return _run_module_code(code, init_globals, run_name, path_name)
  File "/shared/apps/python/Python-2.7.5/INSTALL/lib/python2.7/runpy.py", line 82, in _run_module_code
    mod_name, mod_fname, mod_loader, pkg_name)
  File "/shared/apps/python/Python-2.7.5/INSTALL/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/home/chiyw1992/Code/RMG-Py/rmg.py", line 162, in <module>
    rmg.execute(**kwargs)
  File "/home/chiyw1992/Code/RMG-Py/rmgpy/rmg/main.py", line 732, in execute
    bimolecularReact=self.bimolecularReact)
  File "/home/chiyw1992/Code/RMG-Py/rmgpy/rmg/model.py", line 643, in enlarge
    spcs = [self.retrieveNewSpecies(rxn) for rxn in rxns]
  File "/home/chiyw1992/Code/RMG-Py/rmgpy/rmg/model.py", line 1711, in retrieveNewSpecies
    return self.getSpecies(obj)
  File "/home/chiyw1992/Code/RMG-Py/rmgpy/rmg/model.py", line 1696, in getSpecies
    raise e
KeyError: 7480
```

Is the priority list for pruning based only on reaction system 1? or something else?

I think we should not be pruning something that is the most important species.

**Comment by mjohnson541** [comment]
Pruning is done based on the maximum species flux over all reaction systems.  The only simple explanation I can think of right now is that 1000 of your edge species are ineligible for pruning so it simply dumped every eligible edge species.  But that would imply that library species in edge + recently (default is last 2 iterations) added edge species ~ 1000.  Is that the case here?

**Comment by KEHANG** [comment]
Can you show us all the tolerances you are using in the input?

**Comment by chiyw13** [comment]
This is my input file:
```
database(
    thermoLibraries = ['primaryThermoLibrary'],
    reactionLibraries = [],
    seedMechanisms = [],
    kineticsDepositories = ['training'],
    kineticsFamilies = 'default',
    kineticsEstimator = 'rate rules',
)

species(
    label='ethane',
    reactive=True,
    structure=SMILES("CC"),
)
species(
    label='O2',
    reactive=True,
    structure=SMILES("[O][O]"),
)
species(
    label='He',
    reactive=False,
    structure=SMILES("[He]"),
)

simpleReactor(
    temperature=(500,'K'),
    pressure=(1.0,'atm'),
    initialMoleFractions={
        "ethane": 1,
        "O2": 3.5,
        "He": 20
    },
    terminationTime = (5,'s'),
)

simpleReactor(
    temperature=(1000,'K'),
    pressure=(1.0,'atm'),
    initialMoleFractions={
        "ethane": 1,
        "O2": 3.5,
        "He": 20
    },
    terminationTime = (5,'s'),
)

simulator(
    atol=1e-16,
    rtol=1e-8,
)

model(
    toleranceKeepInEdge=0.02,
    toleranceMoveToCore=0.5,
    toleranceInterruptSimulation=1e8,
    maximumEdgeSpecies=1000,
    toleranceMoveEdgeReactionToCore=30.0, 
    toleranceMoveEdgeReactionToSurface=10.0, 
    toleranceMoveSurfaceReactionToCore=20.0, 
    toleranceMoveSurfaceSpeciesToCore=0.02, 
    #maxNumObjsPerIter=1, #
    filterReactions=True,
    minCoreSizeForPrune=30,
    minSpeciesExistIterationsForPrune=2,
)

generatedSpeciesConstraints(
    allowed=['input species','seed mechanisms','reaction libraries'],   
    maximumRadicalElectrons=3,
    allowSingletO2=False,
)
```
Maybe the `maximumEdgeSpecies`  is really too small, so I am running another case with larger  `maximumEdgeSpecies`  now.``

**Comment by KEHANG** [comment]
@chiyw13 larger `maximumEdgeSpecies` should help in this case, usually we choose this parameter to be above `10,000`. I think it'll be beneficial to explain the logical steps of pruning.

First, RMG ranks each edge species by flux ratio (flux/characteristic flux).

Second, collect all the edge species with low ratio (ratio < `toleranceKeepInEdge`), which will be pruned for sure.

Third, if the total number of edge species is still higher than `maximumEdgeSpecies`, which means we are concerned **more about memory than accuracy**. So RMG would collect more important species. In your case, `maximumEdgeSpecies=1000` basically tells RMG that it's allowed to delete anything until `1000` is met. That's why your 7480-indexed species got pruned although it's going to enter core soon later.

But I think your suggestion is still valid, we can add a filter in third step, to make sure really important species (like the ones with flux ratio higher than `toleranceMoveToCore`) not be collected by pruning algorithm. Working on it, would like you to try that once it's done.

**Comment by KEHANG** [comment]
This new branch [`safe_pruning`](https://github.com/ReactionMechanismGenerator/RMG-Py/tree/safe_pruning) has one commit (the final commit) that does the filtering I suggested. Let me know if that can fix the problem. It should be able to handle cases like `maximumEdgeSpecies=1000` now.

**Comment by mliu49** [comment]
This should be fixed by #1390.
